C error when using expr? in an `or {}` block

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.4 5306469.b7b47fe
OS: windows, Microsoft Windows 10 Pro v19045 64-bit
Processor: 20 cpus, 64bit, little endian, 

getwd: C:\Users\user\Desktop\Code\leaflang
vexe: C:\Users\user\Desktop\Code\v\v.exe
vexe mtime: 2024-02-08 11:43:40

vroot: OK, value: C:\Users\user\Desktop\Code\v
VMODULES: OK, value: C:\Users\user\.vmodules
VTMP: OK, value: C:\Users\user\AppData\Local\Temp\v_0

Git version: git version 2.42.0.windows.1
Git vroot status: weekly.2024.06-18-gb7b47fe1
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,

operable program or batch file.


thirdparty/tcc status: thirdparty-windows-amd64 a39eb79b

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg test.v`
```v
fn find_startswith_string(a []string, search string) ?string {
	for s in a {
		if s.starts_with(search) {
			return s
		}
	}
	return none
}

fn find_any_startswith_string(a []string, b []string, search string) ?string {
 	return find_startswith_string(a, search) or { find_startswith_string(b, search)? }
}

fn main() {
	deadbeef := find_any_startswith_string(['foobar', 'barfoo'], ['deadbeef', 'beefdead'],
		'dead')
	assert deadbeef or { panic('unreachable') } == 'deadbeef'
}

```

**What did you expect to see?**

Program compiles, assert succeeds


**What did you see instead?**
```
==================
C:/Users/user/AppData/Local/Temp/v_0/test.01HP4A5Z95NS0Y0TFWD31GY4KF.tmp.c:6862: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/user/AppData/Local/Temp/v_0/test.01HP4A5Z95NS0Y0TFWD31GY4KF.tmp.c:13080: error: cannot convert 'struct string' to 'struct _option_string'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

**Additional info**

This is a regression from https://github.com/vlang/v/pull/20717. This check work will work:
```v
// since we can return options/results from or blocks, `find_startswith_string(b, search)?` is redundant
return find_startswith_string(a, search) or { find_startswith_string(b, search) }
``` 
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -6620,14 +6620,17 @@ fn (mut g Gen) gen_or_block_stmts(cvar_name string, cast_typ string, stmts []ast
 					g.stmt_path_pos.delete_last()
 				} else {
 					mut is_array_fixed := false
+					mut return_wrapped := false
 					if is_option {
 						is_array_fixed = expr_stmt.expr is ast.ArrayInit
 							&& g.table.final_sym(return_type).kind == .array_fixed
 						if !is_array_fixed {
 							if g.inside_return && !g.inside_struct_init
 								&& expr_stmt.expr is ast.CallExpr
-								&& return_type.has_option_or_result() {
+								&& return_type.has_option_or_result()
+								&& expr_stmt.expr.or_block.kind == .absent {
 								g.write('${cvar_name} = ')
+								return_wrapped = true
 							} else {
 								g.write('*(${cast_typ}*) ${cvar_name}.data = ')
 							}
@@ -6653,6 +6656,9 @@ fn (mut g Gen) gen_or_block_stmts(cvar_name string, cast_typ string, stmts []ast
 					}
 					g.writeln(';')
 					g.stmt_path_pos.delete_last()
+					if return_wrapped {
+						g.writeln('return ${cvar_name};')
+					}
 				}
 			}
 		} else {

--- a/vlib/v/tests/option_call_on_orexpr_test.v
+++ b/vlib/v/tests/option_call_on_orexpr_test.v
@@ -13,8 +13,17 @@ fn find_any_startswith_string(a []string, b []string, search string) ?string {
 	return find_startswith_string(a, search) or { find_startswith_string(b, search) }
 }
 
+fn find_any_startswith_string_unwrapped(a []string, b []string, search string) ?string {
+	return find_startswith_string(a, search) or { find_startswith_string(b, search)? }
+}
+
 fn test_main() {
-	deadbeef := find_any_startswith_string(['foobar', 'barfoo'], ['deadbeef', 'beefdead'],
+	var1 := find_any_startswith_string(['foobar', 'barfoo'], ['deadbeef', 'beefdead'],
 		'dead')
-	assert deadbeef or { panic('unreachable') } == 'deadbeef'
+	assert var1 or { panic('unreachable') } == 'deadbeef'
+	dump(var1)
+
+	var2 := find_any_startswith_string_unwrapped(['foobar', 'barfoo'], ['deadbeef', 'beefdead'],
+		'dead')
+	dump(var2)
 }

